lib/opencensus/trace/integrations/rack_middleware.rb

```diff
@@ -15,24 +15,86 @@
       class RackMiddleware
-        def initialize app
+        ##
+        # Create the Rack middleware.
+        #
+        # @param [#call] app Next item on the middleware stack
+        # @param [#export] exporter The exported used to export captured spans
+        #     at the end of the request. Optional: If omitted, uses the exporter
+        #     in the current config.
+        #
+        def initialize app, exporter: nil
           @app = app
+          @exporter = exporter || OpenCensus::Trace::Config.exporter
         end
 
+        ##
+        # Run the Rack middleware.
+        #
+        # @param [Hash] env The rack environment
+        # @return [Array] The rack response. An array with 3 elements: the HTTP
+        #     response code, a Hash of the response headers, and the response
+        #     body which must respond to `each`.
+        #
         def call env
-          OpenCensus::Context.clear
-          trace = OpenCensus::Trace.start
-          begin
-            OpenCensus::Trace.in_span "rack-request" do |_span|
-              @app.call env
+          OpenCensus::Trace.start_request_trace rack_env: env do |span_context|
+            begin
+              span_context.in_span get_path(env) do |span|
+                start_request span, env
+                @app.call(env).tap do |response|
+                  finish_request span, response
+                end
+              end
+            ensure
+              @exporter.export span_context.build_contained_spans
             end
-          ensure
-            send_trace trace, env
           end
         end
 
-        def send_trace trace, env; end
+        private
+
+        def get_path env
+          path = "#{env['SCRIPT_NAME']}#{env['PATH_INFO']}"
+          path = "/#{path}" unless path.start_with? "/"
+          path
+        end
+
+        def get_host env
+          env["HTTP_HOST"] || env["SERVER_NAME"]
+        end
+
+        def get_url env
+          path = get_path env
+          host = get_host env
+          scheme = env["SERVER_PROTOCOL"]
+          query_string = env["QUERY_STRING"].to_s
+          url = "#{scheme}://#{host}#{path}"
+          url = "#{url}?#{query_string}" unless query_string.empty?
+          url
+        end
+
+        def start_request span, env
+          span.put_attribute "/http/host", get_host(env)
+          span.put_attribute "/http/url", get_url(env)
+          span.put_attribute "/http/method", env["REQUEST_METHOD"]
+          span.put_attribute "/http/client_protocol", env["SERVER_PROTOCOL"]
+          span.put_attribute "/http/user_agent", env["HTTP_USER_AGENT"]
+          span.put_attribute "/pid", ::Process.pid.to_s
+          span.put_attribute "/tid", ::Thread.current.object_id.to_s
+        end
+
+        def finish_request span, response
+          if response.is_a?(::Array) && response.size == 3
+            span.set_status response[0]
+            response[1]["Trace-Context"] = span.context.to_trace_context_header
```

Comment:
Also see #29

Reply:
This is a response header. Is Trace-Context defined on responses?

Reply:
I guess it's not. Propagating the `X-Cloud-Trace` header back upstream was part of the Stackdriver Trace header requirements. I can remove it here.

Reply:
Although I do remember conversations about returning the header in the response.

Reply:
I know there have been conversations, but I haven't seen a solid decision or spec on the semantics of the response header. We can check the other language implementations to see if anyone is actually setting a header.

Reply:
Ok, I don't see anyone else except php that's doing it. I'll remove it here.